File: java-sdk/src/main/java/com/cleanster/sdk/client/CleansterClient.java
package com.cleanster.sdk.client;

import com.cleanster.sdk.api.*;

/**
 * Main entry point for the Cleanster Partner API SDK.
 *
 * <p>Usage example:</p>
 * <pre>{@code
 * CleansterClient client = CleansterClient.sandboxClient("your-access-key");
 *
 * // Create a user
 * CreateUserRequest req = new CreateUserRequest();
 * req.setEmail("user@example.com");
 * req.setFirstName("John");
 * req.setLastName("Doe");
 * CreateUserResponse user = client.users().createUser(req);
 *
 * // Fetch the access token and set it for subsequent calls
 * String token = client.users().fetchAccessToken(user.getData().getId()).getData().getToken();
 * client.setAccessToken(token);
 *
 * // List bookings
 * BookingListResponse bookings = client.bookings().getBookings(null, null);
 * }</pre>
 */
public class CleansterClient {

    private final HttpClient httpClient;
    private final BookingApi bookingApi;
    private final UserApi userApi;
    private final PropertyApi propertyApi;
    private final ChecklistApi checklistApi;
    private final OtherApi otherApi;
    private final BlacklistApi blacklistApi;
    private final PaymentMethodApi paymentMethodApi;
    private final WebhookApi webhookApi;

    public CleansterClient(CleansterConfig config) {
        this.httpClient = new HttpClient(config);
        this.bookingApi = new BookingApi(httpClient);
        this.userApi = new UserApi(httpClient);
        this.propertyApi = new PropertyApi(httpClient);
        this.checklistApi = new ChecklistApi(httpClient);
        this.otherApi = new OtherApi(httpClient);
        this.blacklistApi = new BlacklistApi(httpClient);
        this.paymentMethodApi = new PaymentMethodApi(httpClient);
        this.webhookApi = new WebhookApi(httpClient);
    }

    /**
     * Create a client pointing to the sandbox environment.
     *
     * @param accessKey Your Cleanster partner access key
     */
    public static CleansterClient sandboxClient(String accessKey) {
        return new CleansterClient(CleansterConfig.sandboxBuilder(accessKey).build());
    }

    /**
     * Create a client pointing to the production environment.
     *
     * @param accessKey Your Cleanster partner access key
     */
    public static CleansterClient productionClient(String accessKey) {
        return new CleansterClient(CleansterConfig.productionBuilder(accessKey).build());
    }

    /**
     * Set the user-level access token (Bearer token) for authenticated requests.
     * Call this after obtaining a token via {@link UserApi#fetchAccessToken(int)}.
     *
     * @param token The bearer token string
     */
    public void setAccessToken(String token) {
        httpClient.setBearerToken(token);
    }

    /**
     * Get the currently set access token, or null if not set.
     */
    public String getAccessToken() {
        return httpClient.getBearerToken();
    }

    /**
     * Access booking-related API operations (create, list, cancel, reschedule, etc.)
     */
    public BookingApi bookings() {
        return bookingApi;
    }

    /**
     * Access user-related API operations (create user, fetch token, verify JWT)
     */
    public UserApi users() {
        return userApi;
    }

    /**
     * Access property-related API operations (CRUD, cleaners, iCal, checklists)
     */
    public PropertyApi properties() {
        return propertyApi;
    }

    /**
     * Access checklist-related API operations (list, create, update, delete)
     */
    public ChecklistApi checklists() {
        return checklistApi;
    }

    /**
     * Access other utility operations (services, plans, cost estimate, cleaners, coupons)
     */
    public OtherApi other() {
        return otherApi;
    }

    /**
     * Access blacklist management operations
     */
    public BlacklistApi blacklist() {
        return blacklistApi;
    }

    /**
     * Access payment method operations
     */
    public PaymentMethodApi paymentMethods() {
        return paymentMethodApi;
    }

    /**
     * Access webhook management operations
     */
    public WebhookApi webhooks() {
        return webhookApi;
    }
}

File: android-sdk/src/main/kotlin/com/cleanster/android/CleansterClient.kt
package com.cleanster.android

import com.cleanster.android.api.*
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

/**
 * Main entry point for the Cleanster Android SDK (Retrofit).
 *
 * Usage:
 *   val client = CleansterClient.sandbox("your-access-key")
 *   val token  = client.users.fetchAccessToken(userId)
 *   client.setToken(token.data?.token ?: "")
 *   val bookings = client.bookings.getBookings()
 */
class CleansterClient private constructor(
    private val accessKey: String,
    baseUrl: String,
    enableLogging: Boolean = false,
) {
    private var userToken: String? = null

    private val gson: Gson = GsonBuilder().create()

    private val httpClient: OkHttpClient = OkHttpClient.Builder()
        .addInterceptor(AuthInterceptor(accessKey) { userToken })
        .also { builder ->
            if (enableLogging) {
                builder.addInterceptor(
                    HttpLoggingInterceptor().apply {
                        level = HttpLoggingInterceptor.Level.BODY
                    }
                )
            }
        }
        .build()

    private val retrofit: Retrofit = Retrofit.Builder()
        .baseUrl(baseUrl)
        .client(httpClient)
        .addConverterFactory(GsonConverterFactory.create(gson))
        .build()

    /** Bookings — schedule, reschedule, cancel, assign cleaners, pay, feedback, chat */
    val bookings = BookingsApi(retrofit)

    /** Users — create accounts and manage JWTs */
    val users = UsersApi(retrofit)

    /** Properties — CRUD, iCal sync, cleaner lists, checklists */
    val properties = PropertiesApi(retrofit)

    /** Checklists — create and manage reusable task lists */
    val checklists = ChecklistsApi(retrofit)

    /** Other — services, plans, cost estimates, extras, available cleaners, coupons */
    val other = OtherApi(retrofit)

    /** Blacklist — block specific cleaners from your properties */
    val blacklist = BlacklistApi(retrofit)

    /** Payment Methods — Stripe and PayPal payment method management */
    val paymentMethods = PaymentMethodsApi(retrofit)

    /** Webhooks — subscribe to booking lifecycle events */
    val webhooks = WebhooksApi(retrofit)

    /** Set the user-scoped JWT token. Call after fetchAccessToken(). */
    fun setToken(token: String) {
        userToken = token
    }

    /** Clear the current user token (e.g. on logout). */
    fun clearToken() {
        userToken = null
    }

    companion object {
        private const val SANDBOX_URL =
            "https://partner-sandbox-dot-official-tidyio-project.ue.r.appspot.com/public/"
        private const val PRODUCTION_URL =
            "https://partner-dot-official-tidyio-project.ue.r.appspot.com/public/"

        /**
         * Create a sandbox client for development and testing.
         * @param accessKey Your partner access key
         * @param enableLogging Set true to log requests/responses (development only)
         */
        fun sandbox(accessKey: String, enableLogging: Boolean = false) =
            CleansterClient(accessKey, SANDBOX_URL, enableLogging)

        /**
         * Create a production client.
         * @param accessKey Your partner access key
         */
        fun production(accessKey: String) =
            CleansterClient(accessKey, PRODUCTION_URL, enableLogging = false)

        /**
         * Create a client with a custom base URL (used in tests with MockWebServer).
         */
        fun custom(accessKey: String, baseUrl: String, enableLogging: Boolean = false) =
            CleansterClient(accessKey, baseUrl, enableLogging)
    }
}
